etna/datasets/tsdataset.py

```diff
@@ -132,15 +133,34 @@ def transform(self, transforms: Sequence["Transform"]):
         self.transforms = transforms
         for transform in self.transforms:
             tslogger.log(f"Transform {transform.__class__.__name__} is applied to dataset")
+            columns_before = set(self.df.columns.get_level_values("feature"))
             self.df = transform.transform(self.df)
+            columns_after = set(self.df.columns.get_level_values("feature"))
+            self._update_regressors(transform=transform, columns_before=columns_before, columns_after=columns_after)
 
     def fit_transform(self, transforms: Sequence["Transform"]):
         """Fit and apply given transforms to the data."""
         self._check_endings()
         self.transforms = transforms
         for transform in self.transforms:
             tslogger.log(f"Transform {transform.__class__.__name__} is applied to dataset")
+            columns_before = set(self.df.columns.get_level_values("feature"))
             self.df = transform.fit_transform(self.df)
+            columns_after = set(self.df.columns.get_level_values("feature"))
+            self._update_regressors(transform=transform, columns_before=columns_before, columns_after=columns_after)
+
+    def _update_regressors(self, transform: "Transform", columns_before: Set[str], columns_after: Set[str]):
+        from etna.transforms.base import FutureMixin
+
+        new_columns = list(columns_after - columns_before)
+        if isinstance(transform, FutureMixin):
+            self._regressors.extend(new_columns)
+        elif hasattr(transform, "in_column"):
+            in_columns = transform.in_column if isinstance(transform.in_column, list) else [transform.in_column]  # type: ignore
+            new_columns = [new_column for i, new_column in enumerate(new_columns) if in_columns[i] in self.regressors]
```

Comment:
```suggestion
            new_columns = [new_column for new_column in new_columns if new_column in self.regressors]
```

Reply:
This is not equivalent to the previous version

Reply:
Sorry, to many equvalent names.

Reply:
May be make code more explicit: we enumerate over on iterator and use index for another -- let's make it more explicit.